edb/edgeql/compiler/normalization.py

```diff
@@ -0,0 +1,539 @@
+        ignorenames=ignorenames,
+    )
+
+    return node
+
+
+@normalize.register
+def normalize_GroupQuery(
+    node: qlast.GroupQuery,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> qlast.GroupQuery:
+    raise ValueError(f'normalize: cannot handle {node!r}')
+
+
+@normalize.register
+def compile_Path(
+    node: qlast.Path,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> qlast.Path:
+
+    node = copy.copy(node)
+
+    steps = []
+    for step in node.steps:
+        if isinstance(step, (qlast.Expr, qlast.TypeIntersection)):
+            steps.append(
+                normalize(
+                    step,
+                    schema=schema,
+                    modaliases=modaliases,
+                    ignorenames=ignorenames,
+                )
+            )
+        elif isinstance(step, qlast.ObjectRef):
+            # This is a specific path root, resolve it.
+            if not step.module and step.name not in ignorenames:
+                obj = schema.get(
+                    step.name,
+                    default=None,
+                    module_aliases=modaliases,
+                )
+                if obj is not None:
+                    step.module = obj.get_name(schema).module
+                elif None in modaliases:
+                    # Even if the name was not resolved in the
+                    # schema it may be the name of the object
+                    # being defined, as such the default module
+                    # should be used. Names that must be ignored
+                    # (like aliases and parameters) have already
+                    # been filtered by the ignorenames.
+                    step.module = modaliases[None]
+            steps.append(step)
+        else:
+            steps.append(step)
+
+    node.steps = steps
+
+    return node
+
+
+@normalize.register
+def compile_FunctionCall(
+    node: qlast.FunctionCall,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> qlast.FunctionCall:
+
+    node = copy.copy(node)
+
+    if isinstance(node.func, str) and node.func not in ignorenames:
+        funcs = schema.get_functions(
+            node.func, default=tuple(), module_aliases=modaliases)
+        if funcs:
+            # As long as we found some functions, they will be from
+            # the same module (the first valid resolved module for the
+            # function name will mask "std").
+            _, fname = funcs[0].get_name(schema).as_tuple()
+            _, module, name = sn.split_name(sn.shortname_from_fullname(fname))
+            node.func = (module, name)
+        elif None in modaliases:
+            # Even if the name was not resolved in the schema it may
+            # be the name of the object being defined, as such the
```

Comment:
What do you mean by "the name of the object being defined"?  A body of a recursive function?

Reply:
More like hand-written DDL for recursive types:
```
CREATE TYPE Foo {
    CREATE MULTI LINK children -> Foo;
}
```
Basically when this is not coming from schema, but is created by hand, the command can easily be appearing like that instead of being separated into `CREATE`/`ALTER`.